server/src/challenge/challenge.service.ts

```diff
@@ -132,4 +139,39 @@ export class ChallengeService {
 
     return eventTracker;
   }
+
+  /** Check if the current event can return rewards */
+  async checkForReward(user: User, eventBase: EventBase, eventTracker: EventTracker) {
+    const userRewards = await user.rewards.loadItems();
+
+    //If User has not completed the event/done all the challenges then:
+    if(eventTracker.completed.count() !== eventBase.challengeCount){
+      return userRewards
+    }
+
+    const rewardType = eventBase.rewardType
+    if(rewardType === EventRewardType.LIMITED_TIME_EVENT && eventBase.time > new Date()){
+      return userRewards
+    }
+    
+    if(rewardType === EventRewardType.PERPETUAL){
+      const newReward = await this.rewardRepository.findOne({event: eventBase});
+      if(newReward !== null){
+        newReward.claimingUser = user
+        newReward.isRedeemed = false
+        await this.rewardRepository.persistAndFlush(newReward);
+        this.clientService.emitUpdateRewardData(user, newReward)
```

Comment:
Make sure to use the DTO when sending data to the user (you would not be sending newReward to the user).